### PR TITLE
ARROW-13113: [R] use RTasks to manage parallel in converting arrow to R

### PR DESCRIPTION
**Pull request by romainfrancois**

### CODE REVIEW

**Comment by romainfrancois** [comment]
also deals with: 
 - https://issues.apache.org/jira/browse/ARROW-12939, closes #10444
 - https://issues.apache.org/jira/browse/ARROW-13114

**Comment by romainfrancois** [comment]
refactoring the `Table$create()`and `RecordBatch` internals to take advantage of altrep when a column is fit for altrep (i.e. at the moment is a double array with no nulls, or an int32 array with no nulls). 

Previously altrep was only triggered when converting the arrays. 

This is not done in parallel because altrep allocates, but also because at least for these, it's cheap/instant.

File: r/src/r_task_group.h [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
```
Comment (nealrichardson) [inline r/src/r_task_group.h:19]: I think this needs the if defined ARROW_R_WITH_ARROW wrapping

**Comment by nealrichardson** [comment]
@ursabot please benchmark lang=R

**Comment by nealrichardson** [comment]
@github-actions crossbow submit -g r

**Comment by romainfrancois** [comment]
I have code that deals with https://issues.apache.org/jira/browse/ARROW-13164 on top of this pull request. Can/Should I merge it in here or can this be merged and then I rebase ? @nealrichardson

**Comment by nealrichardson** [comment]
> I have code that deals with https://issues.apache.org/jira/browse/ARROW-13164 on top of this pull request. Can/Should I merge it in here or can this be merged and then I rebase ? @nealrichardson

I think this can be merged and then you rebase for ARROW-13164, but I'll defer to @jonkeane to give the final 👍 and merge here, I've been out all week and haven't been keeping up with everything that's going on.

**Comment by romainfrancois** [comment]
Thanks @nealrichardson. @jonkeane did you have a chance to review this ?

**Comment by jonkeane** [comment]
This looks good to merge. I did noticed there is one benchmark that seemed to have gotten longer (https://conbench.ursa.dev/compare/benchmarks/8cd9415be73542b5b6422f112b82ab6c...7752b1285dea4eb6b4af79a42a191189/). It's possible that won't persist, but if it does we can dig into it and address it in a follow on.

**Pull request closed without merging**